python/nx-cugraph/nx_cugraph/utils/misc.py

```diff
@@ -12,43 +12,60 @@
 # limitations under the License.
 from __future__ import annotations
 
+import itertools
 import operator as op
 import sys
 from random import Random
 
 import cupy as cp
 
+try:
+    from itertools import pairwise  # Python >=3.10
+except ImportError:
+
+    def pairwise(it):
+        it = iter(it)
+        for prev in it:
+            for cur in it:
+                yield (prev, cur)
+                prev = cur
+
+
 __all__ = ["_groupby", "_seed_to_int"]
 
 
-def _groupby(groups: cp.ndarray, values: cp.ndarray) -> dict[int, cp.ndarray]:
+def _groupby(
+    groups: cp.ndarray, values: cp.ndarray, groups_are_canonical: bool = False
+) -> dict[int, cp.ndarray]:
```

Comment:
I'm really happy with this groupby implementation btw. Using `prepend` with `cp.diff` is a little slow compared to `cp.diff` w/o it, but this isn't where the bulk of the time is spent.